cou/steps/plan.py

```diff
@@ -391,17 +391,60 @@ def _get_pre_upgrade_steps(analysis_result: Analysis, args: CLIargs) -> list[Pre
             )
         )
 
-    # Add a pre-upgrade step to archive old database data.
-    # This is a performance optimisation.
+    steps.extend(_get_archive_data_steps(analysis_result, args))
+    steps.extend(_get_purge_data_steps(analysis_result, args))
+    return steps
+
+
+def _get_purge_data_steps(analysis_result: Analysis, args: CLIargs) -> list[PreUpgradeStep]:
+    """Get purge nova db data step.
+
+    :param analysis_result: Analysis result
+    :type analysis_result: Analysis
+    :param args: CLI arguments
+    :type args: CLIargs
+    :return: List of post-upgrade steps.
+    :rtype: list[PreUpgradeStep]
+    """
+    if args.purge and any(
+        app.charm == "nova-cloud-controller" and "purge-data" in app.actions
```

Comment:
The purpose here is to check the juju action is exists. It's a new action we backport on previous pulse. Not checking if nova-cloud-controller app is exists.

Reply:
> Please take some inspiration from

Their purpose are different, we only need to run on leader unit.

Reply:
> open a low priority ticket to investigate refactoring to support the case where there are 0 or >1 nova-cloud-controller applications

I don't think it's exists, the database is created and maintain by nova-cloud-controller.

Reply:
> The purpose here is to check the juju action is exists. It's a new action we backport on previous pulse. Not to check if nova-cloud-controller app is exists.

That is not what the unit test [test_get_purge_data_steps_no_app](https://github.com/canonical/charmed-openstack-upgrader/pull/444/files#diff-179d49568202e72cba8c3edef5135987b4013b411d110462ac772d0c4f1dc41b) shows in my point of view, .

> Their purpose are different, we only need to run on leader unit.

Well, then the implementation is missing that because you are using [_get_nova_cloud_controller_unit_name](https://github.com/jneo8/charmed-openstack-upgrader/blob/16fd6608fce1ebb1fb137b412d5455b801714a75/cou/steps/nova_cloud_controller.py#L105) which is taking the first unit which not necessarily is the leader. That information is important and should be added.

Reply:
Sorry, maybe my previous message is not precise. It's ok to run on any unit because mysql-router will route the connection to the mysql leader unit.

Reply:
Resolve because the case is not exists.